blockchain/rolling_merkle.go

```diff
@@ -0,0 +1,149 @@
+package blockchain
+
+import (
+	"crypto/sha256"
+	"math/bits"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// parentHash returns the hash of the left and right hashes passed in.
+func parentHash(l, r chainhash.Hash) chainhash.Hash {
+	h := sha256.New()
+	h.Write(l[:])
+	h.Write(r[:])
+
+	buf := make([]byte, 0, 32)
+	first := h.Sum(buf)
+	h.Reset()
+	h.Write(first)
+	return *((*chainhash.Hash)(h.Sum(buf)))
+}
+
+// rollingMerkleTreeStore calculates the merkle root by only allocating O(logN)
+// memory where N is the total amount of leaves being included in the tree.
+type rollingMerkleTreeStore struct {
+	// roots are where the temporary merkle roots get stored while the
+	// merkle root is being calculated.
+	roots []chainhash.Hash
+
+	// numLeaves is the total leaves the store has processed.  numLeaves
+	// is required for the root calculation algorithm to work.
+	numLeaves uint64
+}
+
+// newRollingMerkleTreeStore returns a rollingMerkleTreeStore with the roots
+// allocated based on the passed in size.
+func newRollingMerkleTreeStore(size int) rollingMerkleTreeStore {
+	var alloc int
+	if size != 0 {
+		alloc = bits.Len64(uint64(size) - 1)
+	}
+	return rollingMerkleTreeStore{roots: make([]chainhash.Hash, 0, alloc)}
+}
+
+// add adds a single hash to the merkle tree store.  Refer to algorithm 1 "AddOne" in
+// the utreexo paper (https://eprint.iacr.org/2019/611.pdf) for the exact algorithm.
+func (s *rollingMerkleTreeStore) add(add chainhash.Hash) {
+	// We can tell where the roots are by looking at the binary representation
+	// of the numLeaves. Wherever there's a 1, there's a root.
+	//
+	// numLeaves of 8 will be '1000' in binary, so there will be one root at
+	// row 3. numLeaves of 3 will be '11' in binary, so there's two roots. One at
+	// row 0 and one at row 1.
+	//
+	// In this loop below, we're looking for these roots by checking if there's
+	// a '1'. If there is a '1', we'll hash the root being added with that root
+	// until we hit a '0'.
+	newRoot := add
+	for h := uint8(0); (s.numLeaves>>h)&1 == 1; h++ {
+		// Grab the last root.
+		root := s.roots[len(s.roots)-1]
+
+		// Pop off the last root.
+		s.roots = s.roots[:len(s.roots)-1]
+
+		// Calculate the hash of the new root and append it.
+		newRoot = parentHash(root, newRoot)
+	}
+	s.roots = append(s.roots, newRoot)
+	s.numLeaves++
+}
+
+// calcMerkleRoot returns the merkle root for the passed in transactions.
+func (s *rollingMerkleTreeStore) calcMerkleRoot(adds []*btcutil.Tx, witness bool) chainhash.Hash {
```

Comment:
Also need unit tests.

Reply:
`CalcMerkleRoot` test is essentially providing unit tests for this as well since `CalcMerkleRoot` is just:

```
s := newRollingMerkleTreeStore(uint64(len(transactions)))                                                                                                                                                                                                                                        
return s.calcMerkleRoot(transactions, witness) 
```

I could add more unit tests to `CalcMerkleRoot`

Reply:
This API kind of makes me uncomfortable. You can use any rollingMerkleTreeStore as a call target but this only really works if that store was the unit store value or if it has only been seeded with Tx's thus far. I'd strongly recommend constructing the store in this function, or changing the adds vector to be the same as the type of `add` for the store.

Reply:
I won't be addressing this in this PR. There's different ways to do a merkle root calculation and the one implemented here should only be used for calculating merkle root hashes for bitcoin header commitment.

It's a mistake for this to be widely available.